src/metax_api/tests/api/rest/base/views/datasets/write.py

```diff
@@ -178,6 +176,109 @@ def test_change_state_field_through_API(self):
         self.assertEqual(response.status_code, status.HTTP_200_OK, response.data)
         self.assertFalse(response.data['state'] == 'changed value')
 
+    def _set_cr_owner_and_state(self, cr_id, state, owner):
+        ''' helper method for testing user accessibility for draft datasets '''
+        cr = CatalogRecord.objects.get(pk=cr_id)
+        cr.state = state
+        cr.user_created = owner
+        cr.metadata_provider_user = owner
+        cr.editor = None # pretend the record was created by user directly
+        cr.force_save()
+
+    @responses.activate
+    def test_endusers_access_to_draft_datasets(self):
+        ''' End user should get published data and his/her drafts '''
+
+        # Set end user
+        self.token = get_test_oidc_token(new_proxy=True)
+        self._use_http_authorization(method='bearer', token=self.token)
+        self._mock_token_validation_succeeds()
+
+        # Create published record with testuser- owner and draft records
+        # with default & testuser-owner
+        self._set_cr_owner_and_state(1, 'published', self.token['CSCUserName']) # Published dataset
+        self.assertEqual(CatalogRecord.objects.get(pk=1).metadata_provider_user, 'testuser')
+
+        self._set_cr_owner_and_state(2, 'draft', self.token['CSCUserName']) # testusers' draft
+        self.assertEqual(CatalogRecord.objects.get(pk=2).metadata_provider_user, 'testuser')
+
+        self._set_cr_owner_and_state(3, 'draft', '#### Some owner who is not you ####') # Draft dataset for some user
+        self.assertNotEqual(CatalogRecord.objects.get(pk=3).metadata_provider_user, 'testuser')
```

Comment:
This block seems to repeat in many tests. It may be a good idea to override the ``setUp()`` method in this class, and call the above lines there to avoid repetition in each test. Remember to call ``super().setUp()`` at the beginning of the overrided method to guarantee the original setting up actions are executed first before your added customizations.

For examples, see class ``CatalogRecordApiWritePreservationStateTests`` how it overrides the ``setUp()`` method and customizes it.